### PR TITLE
Add API Categories Feature

### PR DESCRIPTION
**Pull request by msm1992**
This PR adds API Categories feature.

API categories can be used to group APIs. In earlier versions this was possible by using tagwise groups. With the new addition of API categories feature tagwise grouping feature is now deprecated and the users are advised to use API categories instead. 

Find the feature doc at https://docs.google.com/document/d/1UvLZ5v29q7wjf4qNIa3UrCSIeFeqiNrTmYxSnwNa62o/edit?usp=sharing

Git Issue : https://github.com/wso2/product-apim/issues/7247

### CODE REVIEW

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIAdminImpl.java [hunk 1]
```diff
@@ -280,4 +287,70 @@ public void addBotDataEmailConfiguration(String email) throws APIManagementExcep
...
+
+    public List<APICategory> getAllAPICategoriesOfTenantForAdminListing(String username) throws APIManagementException{
+        int tenantID = APIUtil.getTenantId(username);
+        List<APICategory> categories = getAllAPICategoriesOfTenant(tenantID);
+        if (categories.size() > 0) {
+            for (APICategory category : categories) {
+                int length = isCategoryAttached(category, username);
+                category.setNumberOfAPIs(length);
+            }
+
```
Comment (ruks) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIAdminImpl.java]: extra space

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIAdminImpl.java [hunk 2]
```diff
@@ -280,4 +287,70 @@ public void addBotDataEmailConfiguration(String email) throws APIManagementExcep
...
+    }
+
+    public APICategory getAPICategoryByID(String apiCategoryId) throws APIManagementException {
+        APICategory apiCategory = ApiMgtDAO.getInstance().getAPICategoryByID(apiCategoryId);
+        if (apiCategory != null) {
+            return apiCategory;
+        }else {
+            String msg = "Failed to get APICategory. API category corresponding to UUID " + apiCategoryId
+                    + " does not exist";
+            log.error(msg);
```
Comment (ruks) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIAdminImpl.java]: no need both logging and throwing exception

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/ApiMgtDAO.java [hunk 3]
```diff
@@ -14321,4 +14322,167 @@ public void removeExpiredJWTs() throws APIManagementException {
...
+                category.setName(name);
+                category.setDescription(description);
+                category.setTenantID(tenantID);
+
+                categoriesList.add(category);
+            }
+        } catch (SQLException e) {
+            handleException("Failed to retrieve API categories for tenant " + tenantID, e);
+        }
+
```
Comment (ruks) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/ApiMgtDAO.java]: extra spaces in this method

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIAdminImpl.java (hunk 1 above)
Comment (msm1992) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIAdminImpl.java]: fixed in https://github.com/wso2/carbon-apimgt/pull/8077/commits/e646d5a8cbf82d6fb2ba2cd090e252a4dfaca540

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIAdminImpl.java (hunk 2 above)
Comment (msm1992) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIAdminImpl.java]: fixed in https://github.com/wso2/carbon-apimgt/pull/8077/commits/e646d5a8cbf82d6fb2ba2cd090e252a4dfaca540

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/ApiMgtDAO.java (hunk 3 above)
Comment (msm1992) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/ApiMgtDAO.java]: fixed in https://github.com/wso2/carbon-apimgt/pull/8077/commits/e646d5a8cbf82d6fb2ba2cd090e252a4dfaca540

File: features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/admin/site/blocks/api-category/api-category-add/ajax/api-category-edit.jag [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+
+(function () {
+    response.contentType = "application/json; charset=UTF-8";
+    var mod, result, obj, permissionResult,
+    action = request.getParameter("action"),
+    uuid = request.getParameter("uuid"),
+    site = require("/site/conf/site.json"),
+    msg = require("/site/conf/ui-messages.jag");
+
+    if(uuid == null){
```
Comment (mushthaq33) [inline features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/admin/site/blocks/api-category/api-category-add/ajax/api-category-edit.jag]: Formatting issue

File: features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/admin/site/blocks/api-category/api-category-add/ajax/api-category-edit.jag [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+        uuid = -1;
+    }
+    if(jagg.getUser() == null){
```
Comment (mushthaq33) [inline features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/admin/site/blocks/api-category/api-category-add/ajax/api-category-edit.jag]: Formatting issue

File: features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/admin/site/blocks/api-category/api-category-list/ajax/api-category-list.jag [hunk 6]
```diff
@@ -0,0 +1,36 @@
+<%
+include("/jagg/jagg.jag");
+
+(function () {
+    response.contentType = "application/json; charset=UTF-8";
+    var mod, result, obj,
+    action = request.getParameter("action"),
+    site = require("/site/conf/site.json"),
+    msg = require("/site/conf/ui-messages.jag");
+    if(jagg.getUser() == null){
```
Comment (mushthaq33) [inline features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/admin/site/blocks/api-category/api-category-list/ajax/api-category-list.jag]: Formatting issue

File: features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/admin/site/themes/wso2/templates/api-category/api-category-add/template.jag [hunk 7]
```diff
@@ -0,0 +1,97 @@
+<% jagg.template("api-category/api-category-add", function(inputs, outputs, jagg) { %>
+
+    <%
+    var uuid = request.getParameter("uuid");
+    var action = request.getParameter("action");
+    var pageTitle = i18n.localize("Add API Category");
+    var nameFieldDisableStatus = "";
+
+    if(action == null){
```
Comment (mushthaq33) [inline features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/admin/site/themes/wso2/templates/api-category/api-category-add/template.jag]: Formatting issue. There are other places as well. Please fix them all.

Comment (msm1992) [reply features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/admin/site/themes/wso2/templates/api-category/api-category-add/template.jag]: fixed in https://github.com/wso2/carbon-apimgt/pull/8077/commits/c6c0ee00e2fc4d641567d19b696b59a2a92f077b

File: features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/admin/site/blocks/api-category/api-category-list/ajax/api-category-list.jag (hunk 6 above)
Comment (msm1992) [reply features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/admin/site/blocks/api-category/api-category-list/ajax/api-category-list.jag]: fixed in https://github.com/wso2/carbon-apimgt/pull/8077/commits/c6c0ee00e2fc4d641567d19b696b59a2a92f077b

File: features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/admin/site/blocks/api-category/api-category-add/ajax/api-category-edit.jag (hunk 5 above)
Comment (msm1992) [reply features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/admin/site/blocks/api-category/api-category-add/ajax/api-category-edit.jag]: fixed in https://github.com/wso2/carbon-apimgt/pull/8077/commits/c6c0ee00e2fc4d641567d19b696b59a2a92f077b

File: features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/admin/site/blocks/api-category/api-category-add/ajax/api-category-edit.jag (hunk 4 above)
Comment (msm1992) [reply features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/admin/site/blocks/api-category/api-category-add/ajax/api-category-edit.jag]: fixed in https://github.com/wso2/carbon-apimgt/pull/8077/commits/c6c0ee00e2fc4d641567d19b696b59a2a92f077b

File: features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/admin/modules/api-category/api-category-add.jag [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+        };
+    } catch (e) {
+        log.error(e.message);
+        return {
+            error:e
+        };
+    }
+}
+
+var constructCategoryObject  = function(categoryId, categoryName, description){
```
Comment (mushthaq33) [inline features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/admin/modules/api-category/api-category-add.jag]: Formatting issue

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.admin/src/main/java/org/wso2/carbon/apimgt/rest/api/admin/impl/ApiCategoriesApiServiceImpl.java [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+                    APICategoryMappingUtil.fromCategoryListToCategoryListDTO(categoryList);
+            return Response.ok().entity(categoryListDTO).build();
+        } catch (APIManagementException e) {
+            String errorMessage = "Error while retrieving API categories";
+            RestApiUtil.handleInternalServerError(errorMessage, e, log);
+        }
+        return null;
+    }
+
+    public Response apiCategoriesPost(APICategoryDTO body){
```
Comment (mushthaq33) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.admin/src/main/java/org/wso2/carbon/apimgt/rest/api/admin/impl/ApiCategoriesApiServiceImpl.java]: Formatting issue

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.admin/src/main/java/org/wso2/carbon/apimgt/rest/api/admin/impl/ApiCategoriesApiServiceImpl.java [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.List;
+import javax.ws.rs.core.Response;
+
+public class ApiCategoriesApiServiceImpl extends ApiCategoriesApiService {
+    private static final Log log = LogFactory.getLog(ApiCategoriesApiServiceImpl.class);
+
+    public Response apiCategoriesGet(){
```
Comment (mushthaq33) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.admin/src/main/java/org/wso2/carbon/apimgt/rest/api/admin/impl/ApiCategoriesApiServiceImpl.java]: Formatting issue

File: features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/admin/site/themes/wso2/templates/api-category/api-category-list/js/manage.js [hunk 11]
```diff
@@ -0,0 +1,23 @@
+var deleteAPICategory = function (uuid) {
+    $("#messageModal div.modal-footer").html("");
+    jagg.message({
+        content:i18n.t('Are you sure you want to delete this API Category?'),
+        title:i18n.t('Confirm Deletion'),
+        type:'confirm',
+        anotherDialog:true,
+        okCallback:function(){
```
Comment (mushthaq33) [inline features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/admin/site/themes/wso2/templates/api-category/api-category-list/js/manage.js]: Formatting issue

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIAdminImpl.java [hunk 12]
```diff
@@ -280,4 +287,68 @@ public void addBotDataEmailConfiguration(String email) throws APIManagementExcep
...
+    }
+
+    public void updateCategory(APICategory apiCategory) throws APIManagementException {
+        apiMgtDAO.updateCategory(apiCategory);
+    }
+
+    public void deleteCategory(String categoryID, String username) throws APIManagementException {
+        APICategory category = getAPICategoryByID(categoryID);
+        int attchedAPICount = isCategoryAttached(category, username);
+        if ( attchedAPICount > 0) {
```
Comment (mushthaq33) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIAdminImpl.java]: There is an additional space

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.admin/src/gen/java/org/wso2/carbon/apimgt/rest/api/admin/ApiCategoriesApi.java [hunk 13]
```diff
@@ -0,0 +1,90 @@
+package org.wso2.carbon.apimgt.rest.api.admin;
+
+import org.wso2.carbon.apimgt.rest.api.admin.dto.*;
```
Comment (mushthaq33) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.admin/src/gen/java/org/wso2/carbon/apimgt/rest/api/admin/ApiCategoriesApi.java]: Shall we remove the * imports

File: features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/admin/modules/api-category/api-category-add.jag (hunk 8 above)
Comment (msm1992) [reply features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/admin/modules/api-category/api-category-add.jag]: fixed

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.admin/src/main/java/org/wso2/carbon/apimgt/rest/api/admin/impl/ApiCategoriesApiServiceImpl.java (hunk 9 above)
Comment (msm1992) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.admin/src/main/java/org/wso2/carbon/apimgt/rest/api/admin/impl/ApiCategoriesApiServiceImpl.java]: fixed

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.admin/src/main/java/org/wso2/carbon/apimgt/rest/api/admin/impl/ApiCategoriesApiServiceImpl.java (hunk 10 above)
Comment (msm1992) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.admin/src/main/java/org/wso2/carbon/apimgt/rest/api/admin/impl/ApiCategoriesApiServiceImpl.java]: fixed

File: features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/admin/site/themes/wso2/templates/api-category/api-category-list/js/manage.js (hunk 11 above)
Comment (msm1992) [reply features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/admin/site/themes/wso2/templates/api-category/api-category-list/js/manage.js]: fixed

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIAdminImpl.java (hunk 12 above)
Comment (msm1992) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIAdminImpl.java]: fixed

**Pull request merged**